pkg/agent/manager/sync.go

```diff
@@ -6,189 +6,148 @@ import (
 	"crypto/elliptic"
 	"crypto/rand"
 	"crypto/x509"
+	"time"
 
-	"github.com/golang/protobuf/proto"
 	"github.com/spiffe/spire/pkg/agent/manager/cache"
 	"github.com/spiffe/spire/pkg/common/bundleutil"
 	"github.com/spiffe/spire/pkg/common/telemetry"
 	"github.com/spiffe/spire/pkg/common/util"
 	"github.com/spiffe/spire/proto/spire/api/node"
 	"github.com/spiffe/spire/proto/spire/common"
-	"github.com/zeebo/errs"
 )
 
+type csrRequest struct {
+	EntryID   string
+	SpiffeID  string
+	ExpiresAt time.Time
+}
+
 // synchronize hits the node api, checks for entries we haven't fetched yet, and fetches them.
 func (m *manager) synchronize(ctx context.Context) (err error) {
-	var regEntries map[string]*common.RegistrationEntry
-	var cEntryRequests = entryRequests{}
-
-	regEntries, _, err = m.fetchUpdates(ctx, nil)
+	update, err := m.fetchUpdates(ctx, nil)
 	if err != nil {
 		return err
 	}
 
-	m.clearStaleCacheEntries(regEntries)
-
-	err = m.checkExpiredCacheEntries(cEntryRequests)
-	if err != nil {
-		return err
-	}
-
-	err = m.checkForNewCacheEntries(regEntries, cEntryRequests)
-	if err != nil {
-		return err
-	}
+	// update the cache and build a list of CSRs that need to be processed
+	// in this interval.
+	//
+	// the values in `update` now belong to the cache. DO NOT MODIFY.
+	var csrs []csrRequest
+	var expiring int
+	m.cache.Update(update, func(entry *common.RegistrationEntry, svid *cache.X509SVID) {
+		var expiresAt time.Time
+		switch {
+		case svid == nil:
+			// no SVID
+		case len(svid.Chain) == 0:
+			// SVID has an empty chain. this is not expected to happen.
```

Comment:
must not we do something here? notify in a log or somehitng than we are getting an svid without chain?

Reply:
sure, i can log.